### PR TITLE
FOUNDATIONS: Additional exception handling for comments

### PR DESCRIPTION
**Pull request by cjdutoit**
This PR separates foundation service code from the work done on PR #76

### CODE REVIEW

**Comment by cacti79** [comment]
Hi, friend please go to tools->options and do this place system directive first and run a cleanup.
image

**Comment by cjdutoit** [comment]
@cacti79 done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs [hunk 1]
```diff
@@ -194,5 +194,50 @@ public async Task ShouldThrowServiceExceptionOnAddIfDatabaseUpdateErrorOccursAnd
...
+        {
+            // given
+            DateTimeOffset randomDateTime =
+                GetRandomDateTimeOffset();
+
+            Comment randomComment = CreateRandomComment(randomDateTime);
+            Comment someComment = randomComment;
+            string randomMessage = GetRandomMessage();
+            string exceptionMessage = randomMessage;
+            var foreignKeyConstraintConflictException =
```
Comment (hassanhabib) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs:209]: Put a new line between multi-liners and one-liners

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs [hunk 2]
```diff
@@ -194,5 +194,50 @@ public async Task ShouldThrowServiceExceptionOnAddIfDatabaseUpdateErrorOccursAnd
...
+            // then
+            await Assert.ThrowsAsync<CommentDependencyValidationException>(() =>
+                addCommentTask.AsTask());
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTimeOffset(),
+                    Times.Once());
+
+            this.loggingBrokerMock.Verify(broker =>
+               broker.LogError(It.Is(SameExceptionAs(expectedCommentValidationException))),
```
Comment (hassanhabib) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs:235]: Put the `expectedCommentValidationException` to the next line. and take a tab to the next, next line. Do the same for the rest please.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs [hunk 3]
```diff
@@ -194,5 +194,50 @@ public async Task ShouldThrowServiceExceptionOnAddIfDatabaseUpdateErrorOccursAnd
...
+
+            var invalidCommentReferenceException =
+                new InvalidCommentReferenceException(foreignKeyConstraintConflictException);
+
+            var expectedCommentValidationException =
+                new CommentDependencyValidationException(invalidCommentReferenceException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTimeOffset())
+                .Throws(foreignKeyConstraintConflictException);
```
Comment (hassanhabib) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs:220]: You need a tab here. Do the same for the rest please.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs (hunk 1 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs:210]: @hassanhabib done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs (hunk 2 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs]: @hassanhabib done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs (hunk 3 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs]: @hassanhabib done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs [hunk 4]
```diff
@@ -194,5 +194,52 @@ public async Task ShouldThrowServiceExceptionOnAddIfDatabaseUpdateErrorOccursAnd
...
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTimeOffset(),
+                    Times.Once());
+
+            this.loggingBrokerMock.Verify(broker =>
+               broker.LogError(It.Is(SameExceptionAs(
+                   expectedCommentValidationException))),
+                        Times.Once);
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs:240]: order is datetime, logging then storage

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs (hunk 4 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs:241]: @ShriHumrudha done

**Comment by cjdutoit** [comment]
@hassanhabib @ShriHumrudha all done now

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs [hunk 5]
```diff
@@ -194,5 +194,52 @@ public async Task ShouldThrowServiceExceptionOnAddIfDatabaseUpdateErrorOccursAnd
...
+
+        [Fact]
+        public async void ShouldThrowValidationExceptionOnAddIfReferenceErrorOccursAndLogItAsync()
+        {
+            // given
+            DateTimeOffset randomDateTime =
+                GetRandomDateTimeOffset();
+
+            Comment randomComment = CreateRandomComment(randomDateTime);
+            Comment someComment = randomComment;
```
Comment (hassanhabib) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs:206]: For creating `someEntity` you don't need to create `randomEntity`

Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs]: @hassanhabib I removed the randomComment

File: Taarafo.Core/Services/Foundations/Comments/CommentService.Exceptions.cs [hunk 6]
```diff
@@ -52,6 +52,13 @@ private async ValueTask<Comment> TryCatch(ReturningCommentFunction returningComm
 
                 throw CreateAndLogDependencyValidationException(alreadyExistsCommentException);
             }
+            catch (ForeignKeyConstraintConflictException foreignKeyConstraintConflictException)
+            {
+                var invalidCourseAttachmentReferenceException =
```
Comment (ShriHumrudha) [inline Taarafo.Core/Services/Foundations/Comments/CommentService.Exceptions.cs:57]: Change variable name

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs [hunk 7]
```diff
@@ -194,5 +194,48 @@ public async Task ShouldThrowServiceExceptionOnAddIfDatabaseUpdateErrorOccursAnd
...
+            // then
+            await Assert.ThrowsAsync<CommentDependencyValidationException>(() =>
+                addCommentTask.AsTask());
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTimeOffset(),
+                    Times.Once());
+
+            this.loggingBrokerMock.Verify(broker =>
+               broker.LogError(It.Is(SameExceptionAs(
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs:232]: Tab sequence is not aligned.

File: Taarafo.Core/Services/Foundations/Comments/CommentService.Exceptions.cs (hunk 6 above)
Comment (cjdutoit) [reply Taarafo.Core/Services/Foundations/Comments/CommentService.Exceptions.cs]: @ShriHumrudha done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs (hunk 7 above)
Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Add.cs]: @ShriHumrudha done

File: Taarafo.Core/Models/Comments/Exceptions/InvalidCommentReferenceException.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿// ---------------------------------------------------------------
+// Copyright (c) Coalition of the Good-Hearted Engineers
+// FREE TO USE TO CONNECT THE WORLD
+// ---------------------------------------------------------------
+using System;
```
Comment (ShriHumrudha) [inline Taarafo.Core/Models/Comments/Exceptions/InvalidCommentReferenceException.cs:5]: new line space required.

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Modify.cs [hunk 9]
```diff
@@ -213,5 +215,52 @@ public async Task ShouldThrowServiceExceptionOnModifyIfServiceErrorOccursAndLogI
             this.storageBrokerMock.VerifyNoOtherCalls();
             this.loggingBrokerMock.VerifyNoOtherCalls();
         }
+
+        [Fact]
+        public async void ShouldThrowValidationExceptionOnModifyIfReferenceErrorOccursAndLogItAsync()
+        {
+            // given
+            int minuteInPast = GetRandomNegativeNumber();
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Modify.cs:223]: The tests scope if just for handling exception so we need not prepare the test data to pass the valdations. Please check the above tests. Also, this test should be second in the order of this file.

Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Modify.cs]: @ShriHumrudha have updated all the tests.

File: Taarafo.Core/Models/Comments/Exceptions/InvalidCommentReferenceException.cs (hunk 8 above)
Comment (cjdutoit) [reply Taarafo.Core/Models/Comments/Exceptions/InvalidCommentReferenceException.cs:6]: @ShriHumrudha done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Modify.cs [hunk 10]
```diff
@@ -168,10 +188,7 @@ public async Task ShouldThrowDependencyValidationExceptionOnModifyIfDatabaseUpda
         public async Task ShouldThrowServiceExceptionOnModifyIfServiceErrorOccursAndLogItAsync()
```
Comment (ShriHumrudha) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Modify.cs:188]: We need to vaidate the storage broker calls - Update is called never.

Comment (cjdutoit) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Comments/CommentServiceTests.Exceptions.Modify.cs:212]: @ShriHumrudha I hope this is correct now.  I have removed the test data not required and added explicit checks to ensure update is never called.

**Review by ShriHumrudha — APPROVED** [review]

**Comment by cjdutoit** [comment]
@hassanhabib can you merge please?

**Pull request merged**